doc/reference/trd-appid.md

```diff
@@ -0,0 +1,396 @@
+of the public key in the header.
+
+The `Rsa3072KeyWithID` value has a data of length of 768 bytes. It contains a public 3072-bit
+RSA key (384 bytes), followed by a 384-byte ciphertext block, consisting of the SHA512 
+hash of the application binary in this Tock binary followed by a 32-bit application
+ID, encrypted by the private key of the public key in the header.
+
+The `Rsa4096KeyWithID` value has a data of length of 1024 bytes. It contains a public 4096-bit
+RSA key (512 bytes), followed by a 512-byte ciphertext block, consisting of the SHA512 
+hash of the application binary in this Tock binary followed by a 32-bit application
+ID, encrypted by the private key of the public key in the header.
+
+
+4 `Verifier` trait
+===============================
+
+The `Verifier` trait defines an interface to a module that accepts,
+passes on, or rejects application credentials. When a Tock board
+asks the kernel to load processes, it passes a reference to a 
+`Verifier`, which the kernel uses to check credentials.
+
+
+```rust
+pub enum VerificationResult {
+  Accept,
+  Pass,
+  Reject
+}
+
+pub trait Verifier {
+  fn require_credential(&self) -> bool;
+  fn check_credentials(&self, 
+                       credentials: &TbfHeaderV2Credentials, 
+                       binary: &mut [u8]) -> VerificationResult;
+}
+```
+
+The kernel, when it loads a Tock binary, scans its headers in order
+from the beginning of the Tock binary. At each
+`TbfHeaderV2Credentials` header it encounters, it calls
+`check_credentials` on the provided `Verifier`. If the `Verifier`
+returns `Accept`, the kernel stops processing credentials and
+continues loading the Tock binary. If the `Verifier` returns `Reject`,
+the kernel stops processing credentials and terminates loading the
+Tock binary. If the `Verifier` returns `Pass`, the kernel tries the
+next `TbfHeaderV2Credentials`, if there is one. 
+
+If the kernel reaches the end of the TBF headers without encountering
+a `Reject` or `Accept` result, it calls `require_credentials` to ask
+the `Verifier` what the default behavior is.  If `require_credentials`
+returns `true`, the kernel rejects the Tock binary and terminates
+loading it. If `require_credentials` returns `false`, the kernel
+accepts the Tock binary and continues loading it. If a Tock binary has
+no `TbfHeaderV2Credentials` headers then there will be no `Accept` or
+`Reject` results and `require_credentials` defines whether to load
+such a binary.
+
+An implementer of `Verifier` sets the security policy of Tock binary
+loading by deciding which types of credentials, and which credentials,
+are acceptable and which are rejected.
+
+If `check_credentials` returns `Accept` for a
+`TbfHeaderV2Credentials`, the kernel stores a reference to this
+`TbfHeaderV2Credentials` in the process structure. This data
+represents the acting credentials of the process.
+
+
+5 Short IDs and the `Compress` trait
```

Comment:
Do `ShortID`s have to be locally unique?

Also I would say that `ShortID`s are compulsory. That is all systems must generate a `ShortID` for an application (with a default implementation of just incrementing)

Reply:
The intention is that both `Verifier` and `Compress` are used by the kernel; the former is used in binary loading, while the latter is used it process starting. So the kernel depends on references to them.

Reply:
I think you missed my first question. Do `ShortID`s have to be locally unique?

Also maybe it should more explicit in the TRD that there must be an implementation of `Compress`

Reply:
Given that in many cases there needs to be a deterministic mapping between credentials and short IDs, I'd conclude that are not assuredly locally unique. E.g., if you have two copies of the identical application binary, they will have the same ShortID.

Reply:
(deleted previous response, in light of core discussion on 9/10)

Reply:
They have to be locally unique, but in the case of global application identifiers also need to be persistent.

Reply:
I'm a little confused what happened here, but sounds good :)